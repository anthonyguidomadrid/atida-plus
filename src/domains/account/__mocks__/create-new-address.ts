export const createNewAddress = {
  redirect: 'some-path',
  salutation: 'string',
  firstName: 'string',
  lastName: 'string',
  address1: 'string',
  address3: 'string',
  zipCode: 'string',
  city: 'string',
  country: 'string',
  iso2Code: 'string',
  phone: 'string',
  isDefaultShipping: true,
  isDefaultBilling: true
}

export const createNewAddressPayload = {
  data: {
    type: 'customers',
    attributes: {
      salutation: 'string',
      firstName: 'string',
      lastName: 'string',
      address1: 'string',
      address3: 'string',
      zipCode: 'string',
      city: 'string',
      country: 'string',
      iso2Code: 'string',
      phone: 'string',
      isDefaultShipping: true,
      isDefaultBilling: true
    }
  }
} as const

export const UpdatedJWTAndRefreshTokens = {
  refreshToken: 'newRefreshToken',
  JWT: 'newJWT',
  expiresIn: '12345'
}

export const sprykerUpdatedJWTAndRefreshTokens = {
  data: {
    type: '',
    id: '',
    attributes: {
      tokenType: 'Bearer',
      expiresIn: '12345',
      accessToken: 'newJWT',
      refreshToken: 'newRefreshToken',
      idCompanyUser: ''
    },
    links: {
      self: ''
    }
  },
  links: {
    self: ''
  }
}
