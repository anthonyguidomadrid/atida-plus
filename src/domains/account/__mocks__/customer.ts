export const createCustomer = {
  salutation: 'Mr',
  firstName: 'Test',
  lastName: 'McTest',
  email: 'some@email.com',
  password: 'somePassword1!',
  confirmPassword: 'somePassword1!',
  acceptedTerms: true,
  emailNotification: true,
  phone: '654654654',
  company: null,
  surcharge: false,
  addresses: {
    addresses: [
      {
        salutation: 'Mr',
        firstName: 'Test',
        lastName: 'McTest',
        address1: 'Brugstraat 123',
        zipCode: '1011 AD',
        city: 'Amsterdam',
        iso2Code: 'PT',
        phone: '088273791298432',
        isDefaultShipping: true,
        isDefaultBilling: true
      }
    ]
  }
}

export const sprykerCustomer = {
  data: {
    type: 'customers',
    id: 'some-id',
    attributes: {
      salutation: 'Mr',
      firstName: 'Test',
      lastName: 'McTest',
      email: 'some@email.com',
      password: 'somePassword1!',
      confirmPassword: 'somePassword1!',
      acceptedTerms: true,
      emailNotification: true,
      company: null,
      surcharge: false
    }
  }
} as const

export const sprykerCustomerWithAddresses = {
  data: {
    ...sprykerCustomer,
    attributes: {
      ...sprykerCustomer.data.attributes
    },
    relationships: {
      addresses: {
        data: [
          { type: 'addresses', id: 'address-id-1' },
          { type: 'addresses', id: 'address-id-2' }
        ]
      }
    }
  },
  included: [
    {
      type: 'addresses',
      id: 'address-id-1',
      attributes: {
        salutation: 'some-salutation',
        firstName: 'some-name',
        lastName: 'some-last-name',
        address1: 'some-address',
        address2: 'some-address2',
        address3: 'some-address3',
        zipCode: 'some-zip-code',
        city: 'some-city',
        country: 'some-country',
        iso2Code: 'some-iso-code',
        company: 'some-company',
        phone: 'some-phone',
        isDefaultShipping: true,
        isDefaultBilling: true,
        isTaxExempt: false
      }
    },
    {
      type: 'addresses',
      id: 'address-id-2',
      attributes: {
        salutation: 'some-other-salutation',
        firstName: 'some-other-name',
        lastName: 'some-other-last-name',
        address1: 'some-other-address',
        address2: 'some-other-address2',
        address3: 'some-other-address3',
        zipCode: 'some-other-zip-code',
        city: 'some-other-city',
        country: 'some-other-country',
        iso2Code: 'some-other-iso-code',
        company: 'some-other-company',
        phone: 'some-other-phone',
        isDefaultShipping: false,
        isDefaultBilling: false,
        isTaxExempt: false
      }
    }
  ]
}
