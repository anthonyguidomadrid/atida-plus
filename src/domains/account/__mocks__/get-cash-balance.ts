export const balance = {
  total: 20
}
