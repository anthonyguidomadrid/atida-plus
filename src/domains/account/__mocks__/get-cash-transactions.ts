import { AtidaCashTransactionHistoryLedgerResponse } from '../types'

export const mockedTransactionHistory = {
  data: [
    {
      transaction_type: 'debit',
      amount: 200,
      transaction_date: '2022-10-04T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 280,
      expiry_date: '2023-10-04T14:20:22.029764445Z',
      transaction_date: '2022-10-04T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 280,
      expiry_date: '2023-07-23T14:20:22.029764445Z',
      transaction_date: '2022-07-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'debit',
      amount: 200,
      transaction_date: '2022-06-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 280,
      expiry_date: '2023-06-23T14:20:22.029764445Z',
      transaction_date: '2022-06-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 500,
      expiry_date: '2023-05-23T14:20:22.029764445Z',
      transaction_date: '2022-05-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'debit',
      amount: 280,
      expiry_date: '2023-04-23T14:20:22.029764445Z',
      transaction_date: '2022-04-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 200,
      expiry_date: '2023-04-23T14:20:22.029764445Z',
      transaction_date: '2022-04-23T14:20:22.029764445Z'
    }
  ],
  metadata: {
    count: 8,
    total: 27
  }
} as AtidaCashTransactionHistoryLedgerResponse

export const mockedTransactionHistoryBayDateArray = {
  transactions: [
    {
      date: '2022-09-30',
      transactions: [
        {
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'debit',
          amount: 200
        },
        {
          amount: 280,
          expiry_date: '2023-09-30T13:36:27.297Z',
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'credit'
        },
        {
          transaction_type: 'debit',
          amount: 200,
          transaction_date: '2022-09-30T13:36:27.297Z'
        },
        {
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'credit',
          amount: 180,
          expiry_date: '2023-09-30T13:36:27.297Z'
        },
        {
          amount: 100,
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'debit'
        },
        {
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'credit',
          amount: 670,
          expiry_date: '2023-09-30T13:36:27.297Z'
        }
      ]
    },
    {
      date: '2022-07-23',
      transactions: [
        {
          transaction_date: '2022-07-23T14:20:22.029764445Z',
          transaction_type: 'debit',
          amount: 200
        },
        {
          expiry_date: '2023-07-23T14:20:22.029764445Z',
          transaction_date: '2022-07-23T14:20:22.029764445Z',
          transaction_type: 'credit',
          amount: 280
        },
        {
          transaction_date: '2022-07-23T14:20:22.029764445Z',
          transaction_type: 'debit',
          amount: 200
        },
        {
          amount: 290,
          expiry_date: '2023-07-23T14:20:22.029764445Z',
          transaction_date: '2022-07-23T14:20:22.029764445Z',
          transaction_type: 'credit'
        }
      ]
    },
    {
      date: '2022-06-23',
      transactions: [
        {
          transaction_type: 'debit',
          amount: 200,
          transaction_date: '2022-06-23T14:20:22.029764445Z'
        },
        {
          expiry_date: '2023-06-23T14:20:22.029764445Z',
          transaction_date: '2022-06-23T14:20:22.029764445Z',
          transaction_type: 'credit',
          amount: 280
        }
      ]
    },
    {
      date: '2022-05-23',
      transactions: [
        {
          transaction_type: 'credit',
          amount: 500,
          expiry_date: '2023-05-23T14:20:22.029764445Z',
          transaction_date: '2022-05-23T14:20:22.029764445Z'
        }
      ]
    },
    {
      date: '2022-04-23',
      transactions: [
        {
          transaction_date: '2022-04-23T14:20:22.029764445Z',
          transaction_type: 'debit',
          amount: 280,
          expiry_date: '2023-04-23T14:20:22.029764445Z'
        },
        {
          transaction_type: 'credit',
          amount: 200,
          expiry_date: '2023-04-23T14:20:22.029764445Z',
          transaction_date: '2022-04-23T14:20:22.029764445Z'
        }
      ]
    }
  ],
  totalPages: 4
}
