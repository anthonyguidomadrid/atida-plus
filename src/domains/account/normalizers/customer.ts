import { removeUndefinedPropertiesFromObject } from '~helpers'
import { Customer, SprykerCustomer } from '../types'
import { ACCOUNT_TYPE_PERSONAL } from '~config/constants/account-types'
import {
  UpdatedJWTAndRefreshTokens,
  SprykerUpdatedJWTAndRefreshTokens
} from '~domains/account/types'
import { parseDateOfBirthFormat } from '../helpers'

export const normalizeCustomer = (
  customer?: SprykerCustomer
): Customer | undefined =>
  customer?.data?.attributes
    ? removeUndefinedPropertiesFromObject({
        salutation: customer.data.attributes.salutation,
        firstName: customer.data.attributes.firstName,
        lastName: customer.data.attributes.lastName,
        dateOfBirth: customer.data.attributes.dateOfBirth
          ? parseDateOfBirthFormat(customer.data.attributes.dateOfBirth, true)
          : undefined,
        magentoCustomerId: customer.data.attributes.magentoCustomerId
          ? customer.data.attributes.magentoCustomerId
          : undefined,
        email: customer.data.attributes.email,
        emailNotification: customer.data.attributes.emailNotification,
        taxReference: customer.data.attributes.taxReference,
        company: customer.data.attributes.company,
        /* There is no default value for this attribute so let's default it to PERSONAL */
        accountType:
          customer.data.attributes.accountType || ACCOUNT_TYPE_PERSONAL,
        surcharge: customer.data.attributes.surcharge,
        receivePersonalRecommendations:
          customer.data.attributes.receivePersonalRecommendations,
        receivePinnedProductsNotifications:
          customer.data.attributes.receivePinnedProductsNotifications,
        invoiceRequired: customer.data.attributes.invoiceRequired,
        phoneNumber: customer.data.attributes.phone,
        createdAt: customer.data.attributes.createdAt
      })
    : undefined

export const normalizeUpdatedJWTAndRefreshTokens = (
  updatedJWTAndRefreshTokens?: SprykerUpdatedJWTAndRefreshTokens
): UpdatedJWTAndRefreshTokens =>
  removeUndefinedPropertiesFromObject({
    refreshToken: updatedJWTAndRefreshTokens?.data?.attributes?.refreshToken,
    JWT: updatedJWTAndRefreshTokens?.data?.attributes?.accessToken,
    expiresIn: updatedJWTAndRefreshTokens?.data?.attributes?.expiresIn
  })
