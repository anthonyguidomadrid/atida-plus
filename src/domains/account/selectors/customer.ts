import { createSelector } from '@reduxjs/toolkit'
import { RootState } from '~domains/redux'

export const selectCustomer = (state: RootState) =>
  state?.client?.account?.['customer']

export const selectIsLoading = createSelector(
  selectCustomer,
  customer => customer?.isLoading ?? false
)

export const selectWasError = createSelector(
  selectCustomer,
  customer => customer?.wasError ?? false
)

export const selectWasSuccess = createSelector(
  selectCustomer,
  customer => customer?.wasSuccess ?? false
)

export const selectError = createSelector(
  selectCustomer,
  customer => customer?.error
)

export const selectCustomerReference = createSelector(
  selectCustomer,
  customer => customer?.reference
)

export const selectMagentoCustomerReference = createSelector(
  selectCustomer,
  customer => customer?.details?.magentoCustomerId
)

export const selectCustomerDetails = createSelector(
  selectCustomer,
  customer => customer?.details
)

export const selectCustomerTaxReference = createSelector(
  selectCustomer,
  customer => customer?.details?.taxReference
)

export const selectCustomerSurcharge = createSelector(
  selectCustomer,
  customer => customer?.details?.surcharge
)

export const selectCustomerFirstName = createSelector(
  selectCustomerDetails,
  details => details?.firstName
)

export const selectHasPreviousSuccessfulOrder = createSelector(
  selectCustomerDetails,
  details => details?.hasPreviousSuccessfulOrder
)

export const selectInvoiceRequired = createSelector(
  selectCustomerDetails,
  details => details?.invoiceRequired
)

export const selectCustomerSalutation = createSelector(
  selectCustomerDetails,
  details => details?.salutation
)

export const selectCustomerEmail = createSelector(
  selectCustomerDetails,
  details => details?.email
)

export const selectCustomerType = createSelector(
  selectCustomerDetails,
  details => details?.accountType
)

export const selectTrimmedCustomerFirstName = createSelector(
  selectCustomerFirstName,
  firstName => {
    const maxNameLength = 15
    const trimNameLength = 12
    return (firstName?.length ?? 0) > maxNameLength
      ? (firstName = firstName?.substring(0, trimNameLength).concat('...'))
      : firstName
  }
)

export const selectIsLoggedIn = createSelector(
  selectCustomerReference,
  reference => !!reference
)

export const selectCustomerShowNotification = createSelector(
  selectCustomer,
  customer => customer?.showNotification
)

export const selectAddresses = createSelector(
  selectCustomer,
  customer => customer?.addresses ?? []
)

export const selectDefaultBillingAddress = createSelector(
  selectAddresses,
  addresses => addresses.find(address => address.isDefaultBilling)
)

export const selectDefaultBillingAddressZipCode = createSelector(
  selectDefaultBillingAddress,
  address => address?.zipCode ?? ''
)

export const selectDefaultShippingAddress = createSelector(
  selectAddresses,
  addresses => addresses.find(address => address.isDefaultShipping)
)

export const selectDefaultShippingAddressZipCode = createSelector(
  selectDefaultShippingAddress,
  address => address?.zipCode ?? ''
)

export const selectDefaultShippingAddressIsTaxExempt = createSelector(
  selectDefaultShippingAddress,
  address => address?.isTaxExempt
)

export const selectDefaultShippingAddressIsoCode = createSelector(
  selectDefaultShippingAddress,
  address => address?.iso2Code
)

export const selectDefaultShippingAddressRegion = createSelector(
  selectDefaultShippingAddress,
  address => address?.province ?? ''
)
