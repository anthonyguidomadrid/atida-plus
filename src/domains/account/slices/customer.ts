import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import type {
  GetCustomerTriggerPayload,
  Customer,
  ReadCustomerPayload,
  CustomerAddress,
  RequestInvoiceTriggerPayload
} from '../types'

type CustomerState = {
  isLoading: boolean
  wasSuccess: boolean
  wasError: boolean
  error?: string
  reference?: string
  details?: Partial<Customer>
  addresses?: CustomerAddress[]
  showNotification: boolean
}

const initialState: CustomerState = {
  isLoading: false,
  wasSuccess: false,
  wasError: false,
  showNotification: false
}

export const customerSlice = createSlice({
  name: 'account/customer',
  initialState,
  reducers: {
    trigger(_state, _action: PayloadAction<GetCustomerTriggerPayload>) {
      // no-op, used for triggering saga
    },
    request(state) {
      state.isLoading = true
      state.wasSuccess = false
      state.wasError = false
      delete state.error
    },
    success(state, action) {
      state.wasSuccess = true
      const { addresses, ...details } = action.payload

      details.hasPreviousSuccessfulOrder =
        state.details?.hasPreviousSuccessfulOrder
      state.details = details
      state.addresses = addresses
    },
    failure(state, action: PayloadAction<{ message: string }>) {
      const { message } = action.payload
      state.wasError = true
      state.error = message
      delete state.reference
      delete state.details
      delete state.addresses
    },
    fulfill(state) {
      state.isLoading = false
    },
    triggerReadCustomer(_state) {
      // no-op, used for triggering saga
    },
    triggerShowNotification(state) {
      state.showNotification = true
    },
    triggerHideNotification(state) {
      state.showNotification = false
    },
    readCustomer(state, action: PayloadAction<ReadCustomerPayload>) {
      state.reference = action.payload.reference
      if (state.details) {
        state.details.hasPreviousSuccessfulOrder =
          action.payload.hasPreviousSuccessfulOrder
      } else {
        state.details = {
          hasPreviousSuccessfulOrder: action.payload.hasPreviousSuccessfulOrder
        }
      }
    },
    invoiceRequireTrigger(
      _state,
      _action: PayloadAction<RequestInvoiceTriggerPayload>
    ) {
      // no-op, used for triggering saga
    },
    invoiceRequireSuccess(state, action) {
      state.wasSuccess = true
      const invoiceRequired = action.payload
      if (state.details) {
        state.details.invoiceRequired = invoiceRequired
      }
    },
    invoiceRequireFailure(state, action) {
      const { message } = action.payload
      state.wasError = true
      state.error = message
    },
    clearData(state) {
      delete state.reference
      delete state.details
      delete state.addresses
    }
  }
})
