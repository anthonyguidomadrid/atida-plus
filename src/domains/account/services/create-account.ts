import getConfig from 'next/config'
import { createClient } from '~helpers'
import type { CreateCustomerRequest, Customer, SprykerCustomer } from '../types'
import { normalizeCustomer } from '../normalizers'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

export const createCustomerAccount = async (
  locale: string,
  data: CreateCustomerRequest,
  anonymousCustomerUniqueId?: string
): Promise<Customer | undefined> => {
  const { serverRuntimeConfig } = getConfig()
  const client = createClient({
    locale,
    options: { baseURL: serverRuntimeConfig.ecommerceHost[locale] },
    addAnonymousCustomerUniqueId: anonymousCustomerUniqueId
  })

  const response = await client.post<SprykerCustomer>('/customers', {
    data: {
      type: 'customers',
      attributes: data
    }
  })

  return normalizeCustomer(response.data)
}
