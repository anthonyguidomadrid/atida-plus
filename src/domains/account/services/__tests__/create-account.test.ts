/**
 * @jest-environment node
 */
import axios from 'axios'
import { createCustomer, sprykerCustomer } from '../../__mocks__/customer'
import { createClient } from '~helpers'
import { createCustomerAccount } from '../create-account'

describe(createCustomerAccount, () => {
  beforeEach(() => {
    ;(axios.post as jest.Mock).mockResolvedValue({ data: sprykerCustomer })
  })

  it('creates the client & passes the locale and anonymous id to spryker request', async () => {
    await createCustomerAccount('en-gb', createCustomer, 'some-unique-id')
    expect(createClient).toHaveBeenCalledWith({
      addAnonymousCustomerUniqueId: 'some-unique-id',
      locale: 'en-gb',
      options: {
        baseURL: 'spryker.co.uk'
      }
    })

    expect(axios.post).toHaveBeenCalledWith(`/customers`, {
      data: {
        attributes: createCustomer,
        type: 'customers'
      }
    })
  })

  it('returns the normalized customer response', async () => {
    const customerResponse = await createCustomerAccount(
      'en-gb',
      createCustomer,
      'some-unique-id'
    )

    expect(customerResponse).toMatchSnapshot()
  })
})
