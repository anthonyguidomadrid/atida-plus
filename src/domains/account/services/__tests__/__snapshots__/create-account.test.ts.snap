// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCustomerAccount returns the normalized customer response 1`] = `
Object {
  "accountType": "Personal",
  "company": null,
  "email": "some@email.com",
  "emailNotification": true,
  "firstName": "Test",
  "lastName": "McTest",
  "salutation": "Mr",
  "surcharge": false,
}
`;
